File: src/StarterGui/LundstrongOrders/cashierGui/setup.client.lua
local path = require(game.Players.LocalPlayer.PlayerGui.LundstrongOrders.cashierGui.filePaths)
local config = require(workspace.LundstrongOrders.Configuration)
local tweenService = game:GetService("TweenService")
local tweeninfo = TweenInfo.new(1)

local function getTableKeys(tab)
    local keyset = {}
    for _,v in ipairs(tab) do
      for k,_ in pairs(v) do
        keyset[#keyset + 1] = k -- More effective/optimised than table.insert(keyset,k)
      end
    end
    return keyset
end

local function deepCopy(original) -- See https://developer.roblox.com/en-us/articles/Cloning-tables for more info
	local copy = {}
	for k, v in pairs(original) do
		if type(v) == "table" then
			v = deepCopy(v)
		end
		copy[k] = v
	end
	return copy
end

local function createSummaryItem(item)
    local clone = path.cashierGui.createOrder.summaryFrame.ScrollingFrame.cloneFrame:Clone()
    clone.Parent = path.cashierGui.createOrder.summaryFrame.ScrollingFrame
    clone.Visible = true
    clone.TextLabel.Text = item.TextLabel.Text
    clone.Name = "itemFrame"
    clone.MouseButton1Up:Connect(function()
        clone:Destroy()
    end)
    clone.TextLabel.MouseButton1Up:Connect(function()
        clone:Destroy()
    end)
end

-- * Cats (Cashier -> Create Order)
for i,v in pairs(getTableKeys(config.MenuItems.MenuItems)) do
    if (i == 1) then
        path.cashierGui.createOrder.Categories.Categories.selectedCategory.usernameText.Text = v
        path.cashierGui.createOrder.Categories.Categories.selectedCategory.LayoutOrder = i
        path.cashierGui.createOrder.Category.usernameText.Text = v
        -- Add items.
        for _,itemText in pairs(config.MenuItems.MenuItems[i][v]) do
            local clone = path.cashierGui.createOrder.ScrollingFrame.cloneFrame:Clone()
            clone.Parent = path.cashierGui.createOrder.ScrollingFrame
            clone.Visible = true
            clone.TextLabel.Text = itemText.ItemName
            clone.Name = "itemFrame"
            clone.MouseButton1Up:Connect(function()
                createSummaryItem(clone)
            end)
            clone.TextLabel.MouseButton1Up:Connect(function()
                createSummaryItem(clone)
            end)
        end
    else
        local clone = path.cashierGui.createOrder.Categories.Categories.selectedCategory:Clone()
        clone.Parent = path.cashierGui.createOrder.Categories.Categories
        clone.usernameText.Text = v
        clone.Name = "Category"
        clone.LayoutOrder = i
    end
end
path.cashierGui.createOrder.Categories.Categories.CanvasSize = UDim2.new(0,0,0, ((#path.cashierGui.createOrder.Categories.Categories:GetChildren() - 1 ) * 40))
if (#path.cashierGui.createOrder.Categories.Categories:GetChildren() < 5) then
    path.cashierGui.createOrder.Categories.Size = UDim2.new(0.192, 0, ((#path.cashierGui.createOrder.Categories.Categories:GetChildren() - 1 ) * 0.068), 0)
end
for _,v in pairs(path.cashierGui.createOrder.Categories.Categories:GetChildren()) do
    if v:IsA("TextButton") then
        if (#path.cashierGui.createOrder.Categories.Categories:GetChildren() < 5) then
            v.Size = UDim2.new(1, 0, 1/(#path.cashierGui.createOrder.Categories.Categories:GetChildren() -1), 0)
        else
            v.Size = UDim2.new(1, -12, 1/(#path.cashierGui.createOrder.Categories.Categories:GetChildren() -1), 0)
        end
    end
end
local function catSelected(newCat)
    -- Bg
    local tween2 = tweenService:Create(path.cashierGui.createOrder.categoryAnim, tweeninfo, { BackgroundTransparency = 1 })
    -- Rotate
    local tween1 = tweenService:Create(path.cashierGui.createOrder.Categories.Categories.selectedCategory.dropdownButton, tweeninfo, { Rotation = 90 })
    path.cashierGui.createOrder.Category.usernameText.Text = newCat.usernameText.Text
    tween1:Play()
    tween1.Completed:Connect(function()
        path.cashierGui.createOrder.Categories.Visible = false
        wait(0.5)
        for _,v in pairs(path.cashierGui.createOrder.ScrollingFrame:GetChildren()) do
            if v:IsA("TextButton") then
                if v.Name == "itemFrame" then
                    v:Destroy()
                end
            end
        end
        for _,itemText in pairs(config.MenuItems.MenuItems[newCat.LayoutOrder][newCat.usernameText.Text]) do
            local clone = path.cashierGui.createOrder.ScrollingFrame.cloneFrame:Clone()
            clone.Parent = path.cashierGui.createOrder.ScrollingFrame
            clone.Visible = true
            clone.TextLabel.Text = itemText.ItemName
            clone.Name = "itemFrame"
            clone.MouseButton1Up:Connect(function()
                createSummaryItem(clone)
            end)
            clone.TextLabel.MouseButton1Up:Connect(function()
                createSummaryItem(clone)
            end)
        end
        tween2:Play()
    end)
end
for _,v in pairs(path.cashierGui.createOrder.Categories.Categories:GetChildren()) do
    if v:IsA("TextButton") then
        v.MouseButton1Up:Connect(function()
            catSelected(v)
        end)
        v.dropdownButton.MouseButton1Up:Connect(function()
            catSelected(v)
        end)
        v.usernameText.MouseButton1Up:Connect(function()
            catSelected(v)
        end)
    end
end
-- * Autofill (Taken from BAE Code)
local Players = game:GetService("Players")
path.cashierGui.usernamePrompt.usernameFrame.usernameText:GetPropertyChangedSignal("Text"):Connect(function()
    if (path.cashierGui.usernamePrompt.usernameFrame.usernameText.Text ~= "") then
        local toReturn = {}
        local Segment = path.cashierGui.usernamePrompt.usernameFrame.usernameText.Text
        local lowerSegment = string.lower(path.cashierGui.usernamePrompt.usernameFrame.usernameText.Text)
        for _,b in next,Players:GetPlayers() do
            if string.sub(b.DisplayName:lower(),1,#Segment) == lowerSegment then
                table.insert(toReturn,b)
            end
        end
        if type(toReturn[1]) == "userdata" then
            path.cashierGui.usernamePrompt.usernameFrame.playerName.Value = toReturn[1].Name
            path.cashierGui.usernamePrompt.usernameFrame.usernamePrompt.Text = toReturn[1].DisplayName
            path.cashierGui.usernamePrompt.usernameFrame.usernameText.Text = string.sub(path.cashierGui.usernamePrompt.usernameFrame.usernamePrompt.Text, 1, string.len(path.cashierGui.usernamePrompt.usernameFrame.usernameText.Text))
        else
            path.cashierGui.usernamePrompt.usernameFrame.usernamePrompt.Text = "" -- No Username
        end
    else
        path.cashierGui.usernamePrompt.usernameFrame.usernamePrompt.Text = "" -- No Username
    end
end)
-- * Cats (Cashier -> View Menu)
for i,v in pairs(getTableKeys(config.MenuItems.MenuItems)) do
    -- Add Div.
    local divClone = path.cashierGui.viewMenu.ScrollingFrame.dividerClone:Clone()
    divClone.Parent = path.cashierGui.viewMenu.ScrollingFrame
    divClone.Name = "divider"
    divClone.TextLabel.Text = v
    divClone.Visible = true
    -- Add items.
    local items = deepCopy(config.MenuItems.MenuItems[i][v])
    while (items[1]) do
        local clone = path.cashierGui.viewMenu.ScrollingFrame.itemGroupClone:Clone()
        clone.Parent = path.cashierGui.viewMenu.ScrollingFrame
        clone.Visible = true
        clone.Name = "itemFrameGroip"
        if (items[4]) then
            clone.Frame4.TextLabel.Text = items[4].ItemName
            table.remove(items, 4)
        else
            clone.Frame4:Destroy()
        end
        if (items[3]) then
            clone.Frame3.TextLabel.Text = items[3].ItemName
            table.remove(items, 3)
        else
            clone.Frame3:Destroy()
        end
        if (items[2]) then
            clone.Frame2.TextLabel.Text = items[2].ItemName
            table.remove(items, 2)
        else
            clone.Frame2:Destroy()
        end
        clone.Frame1.TextLabel.Text = items[1].ItemName
        table.remove(items, 1)
    end
end
-- Username
path.cashierGui.home.username.Text = game.Players.LocalPlayer.DisplayName

File: src/StarterGui/cashierGui/setup.client.ts
import { Players, TweenService } from "@rbxts/services";
import { getConfig } from "ReplicatedStorage/config";
import { cashierGui } from "./cashierGui";

const cashierGuiInstance = script.Parent!.WaitForChild("cashierGui") as cashierGui;
const config = getConfig();

type Category = cashierGui["createOrder"]["Categories"]["Categories"]["cloneCategory"];

// * createOrder
const createSummaryItem = (item: string) => {
	const clone = cashierGuiInstance.createOrder.summaryFrame.ScrollingFrame.cloneFrame.Clone();
	clone.Name = "itemFrame";
	clone.Parent = cashierGuiInstance.createOrder.summaryFrame.ScrollingFrame;
	clone.TextLabel.Text = item;
	clone.Visible = true;
	clone.MouseButton1Up.Connect(() => {
		clone.Destroy();
	});
};

const categorySelected = (newCategory: Category) => {
	const rotateChevron = TweenService.Create(
		cashierGuiInstance.createOrder.Categories.Categories.selectedCategory.dropdownButton,
		new TweenInfo(1),
		{ Rotation: 90 },
	);
	const lightenBackground = TweenService.Create(cashierGuiInstance.createOrder.categoryAnim, new TweenInfo(1), {
		BackgroundTransparency: 1,
	});
	rotateChevron.Completed.Connect(() => {
		cashierGuiInstance.createOrder.Categories.Visible = false;
		// remove items from the old category and add the new ones
		for (const item of cashierGuiInstance.createOrder.ScrollingFrame.GetChildren()) {
			if (item.IsA("TextButton") && item.Name === "itemFrame") {
				item.Destroy();
			}
		}
		for (const item of config.MenuItems[newCategory.Index.Value].Items) {
			const clone = cashierGuiInstance.createOrder.ScrollingFrame.cloneFrame.Clone();
			clone.Name = "itemFrame";
			clone.Parent = cashierGuiInstance.createOrder.ScrollingFrame;
			clone.Visible = true;
			clone.Text = item;
			clone.MouseButton1Up.Connect(() => {
				createSummaryItem(item);
			});
		}

		const hidden = cashierGuiInstance.createOrder.Categories.Categories.selectedCategory.hiddenCategory
			.Value as Category;
		hidden.Visible = true;
		cashierGuiInstance.createOrder.Categories.Categories.selectedCategory.hiddenCategory.Value = newCategory;
		cashierGuiInstance.createOrder.Categories.Categories.selectedCategory.usernameText.Text =
			newCategory.usernameText.Text;
		cashierGuiInstance.createOrder.Category.usernameText.Text = newCategory.usernameText.Text;
		newCategory.Visible = false;

		lightenBackground.Play();
	});
	rotateChevron.Play();
};

// Add Items to scrollingFrame
for (const item of config.MenuItems[0].Items) {
	const clone = cashierGuiInstance.createOrder.ScrollingFrame.cloneFrame.Clone();
	clone.Name = "itemFrame";
	clone.Parent = cashierGuiInstance.createOrder.ScrollingFrame;
	clone.Visible = true;
	clone.Text = item;
	clone.MouseButton1Up.Connect(() => {
		createSummaryItem(item);
	});
}

// Category List (createOrder Frame)
// ? a forEach is used here so we can also get the index
config.MenuItems.forEach((category, index) => {
	const clone = cashierGuiInstance.createOrder.Categories.Categories.cloneCategory.Clone();
	clone.Name = "Category";
	clone.Parent = cashierGuiInstance.createOrder.Categories.Categories;
	clone.usernameText.Text = category.CategoryName;
	clone.Index.Value = index;
	clone.Visible = index === 0 ? false : true;

	if (index === 0) {
		cashierGuiInstance.createOrder.Categories.Categories.selectedCategory.usernameText.Text = category.CategoryName;
		cashierGuiInstance.createOrder.Categories.Categories.selectedCategory.hiddenCategory.Value = clone;
		cashierGuiInstance.createOrder.Category.usernameText.Text = category.CategoryName;
	}

	clone.MouseButton1Up.Connect(() => {
		categorySelected(clone);
	});
	clone.dropdownButton.MouseButton1Up.Connect(() => {
		categorySelected(clone);
	});
	clone.usernameText.MouseButton1Up.Connect(() => {
		categorySelected(clone);
	});
});

if (cashierGuiInstance.createOrder.Categories.Categories.selectedCategory.hiddenCategory.Value) {
	cashierGuiInstance.createOrder.Categories.Categories.selectedCategory.MouseButton1Up.Connect(() => {
		categorySelected(
			cashierGuiInstance.createOrder.Categories.Categories.selectedCategory.hiddenCategory.Value as Category,
		);
	});
	cashierGuiInstance.createOrder.Categories.Categories.selectedCategory.usernameText.MouseButton1Up.Connect(() => {
		categorySelected(
			cashierGuiInstance.createOrder.Categories.Categories.selectedCategory.hiddenCategory.Value as Category,
		);
	});
	cashierGuiInstance.createOrder.Categories.Categories.selectedCategory.dropdownButton.MouseButton1Up.Connect(() => {
		categorySelected(
			cashierGuiInstance.createOrder.Categories.Categories.selectedCategory.hiddenCategory.Value as Category,
		);
	});
}

// * usernamePrompt
cashierGuiInstance.usernamePrompt.usernameFrame.usernameText.GetPropertyChangedSignal("Text").Connect(() => {
	if (cashierGuiInstance.usernamePrompt.usernameFrame.usernameText.Text !== "") {
		const toReturn: Player[] = [];
		const playerInput = cashierGuiInstance.usernamePrompt.usernameFrame.usernameText.Text.lower();
		for (const player of Players.GetPlayers()) {
			if (string.sub(player.Name.lower(), 1, playerInput.size()) === playerInput) {
				toReturn.push(player);
			}
		}
		if (toReturn[0] && toReturn[0].IsA("Player")) {
			cashierGuiInstance.usernamePrompt.usernameFrame.playerName.Value = toReturn[0].Name;
			cashierGuiInstance.usernamePrompt.usernameFrame.usernamePrompt.Text = toReturn[0].Name;
			cashierGuiInstance.usernamePrompt.usernameFrame.usernameText.Text = string.sub(
				toReturn[0].Name,
				1,
				cashierGuiInstance.usernamePrompt.usernameFrame.usernameText.Text.size(),
			);
		} else {
			cashierGuiInstance.usernamePrompt.usernameFrame.usernamePrompt.Text = "";
		}
	} else {
		cashierGuiInstance.usernamePrompt.usernameFrame.usernamePrompt.Text = "";
	}
});
// * viewMenu
for (const category of config.MenuItems) {
	const dividerClone = cashierGuiInstance.viewMenu.ScrollingFrame.dividerClone.Clone();
	dividerClone.Parent = cashierGuiInstance.viewMenu.ScrollingFrame;
	dividerClone.Name = "Divider";
	dividerClone.TextLabel.Text = category.CategoryName;
	dividerClone.Visible = true;

	const itemGroupClone = cashierGuiInstance.viewMenu.ScrollingFrame.itemGroupClone.Clone();
	itemGroupClone.Parent = cashierGuiInstance.viewMenu.ScrollingFrame;
	itemGroupClone.Visible = true;
	itemGroupClone.Name = "itemFrameGroup";

	for (let i = 0; i < category.Items.size(); i = i + 4) {
		// ? + 4 here, as we're doing 4 at a time
		if (category.Items[i + 3] !== undefined) {
			itemGroupClone.Frame4.TextLabel.Text = category.Items[i + 3];
		} else {
			itemGroupClone.Frame4.Destroy();
		}
		if (category.Items[i + 2] !== undefined) {
			itemGroupClone.Frame3.TextLabel.Text = category.Items[i + 2];
		} else {
			itemGroupClone.Frame3.Destroy();
		}
		if (category.Items[i + 1] !== undefined) {
			itemGroupClone.Frame2.TextLabel.Text = category.Items[i + 1];
		} else {
			itemGroupClone.Frame2.Destroy();
		}
		itemGroupClone.Frame1.TextLabel.Text = category.Items[i];
	}
}
// * home
cashierGuiInstance.home.username.Text = Players.LocalPlayer.DisplayName;
